src/types/function.rs

```diff
@@ -39,6 +76,38 @@ impl PyCFunction {
         )
     }
 
+    /// Create a new function from a closure.
+    pub fn new_closure<F>(f: F, py: Python) -> PyResult<&PyCFunction>
+    where
+        F: Fn(&types::PyTuple, Option<&types::PyDict>) -> PyResult<PyObject>,
+    {
+        let boxed_fn: Box<dyn Fn(&types::PyTuple, Option<&types::PyDict>) -> PyResult<PyObject>> =
+            Box::new(f);
+        let function_ptr = Box::into_raw(Box::new(boxed_fn));
+        let capsule_ptr = unsafe {
+            ffi::PyCapsule_New(
+                function_ptr as *mut c_void,
+                closure_capsule_name(),
+                Some(drop_closure),
+            )
+        };
+        let method_def = methods::PyMethodDef::cfunction_with_keywords(
+            "pyo3_closure",
+            methods::PyCFunctionWithKeywords(run_closure),
+            "",
+        );
+        let def = method_def
```

Comment:
I would prefer a new variant for `PyFunctionArguments` and using `new_internal` instead of copying its code here.

Reply:
Right, duplicating the code seems indeed sad, I've made some small changes to move the duplicated bits in a new function. This seemed slightly simpler than adding a variant to `PyFunctionArguments`, especially now that we have generics on `F` at a couple places (my understanding is that we should avoid having a variant case involving pointers in `PyFunctionArguments` but I'm very unsure about this).